presto-main/src/main/java/com/facebook/presto/operator/BigintInMemoryJoinHash.java

```diff
@@ -0,0 +1,232 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.google.common.primitives.Ints;
+import io.airlift.slice.XxHash64;
+import it.unimi.dsi.fastutil.HashCommon;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+
+import java.util.Arrays;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.operator.SyntheticAddress.decodePosition;
+import static com.facebook.presto.operator.SyntheticAddress.decodeSliceIndex;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.util.Objects.requireNonNull;
+
+public final class BigintInMemoryJoinHash
+        implements LookupSource
+{
+    private final LongArrayList addresses;
+    private final PagesHashStrategy pagesHashStrategy;
+
+    private final int channelCount;
+    private final int mask;
+    private final int[] key;
+    private Optional<int[]> positionLinks = Optional.empty(); // lazy initialized
+    private final long totalMemoryUsage;
+    private final long[] values;
+    private Optional<BitSet> nullsMask = Optional.empty(); // lazy initialized
+
+    public BigintInMemoryJoinHash(LongArrayList addresses, PagesHashStrategy pagesHashStrategy, List<List<Block>> channels, List<Integer> joinChannels)
+    {
+        checkState(requireNonNull(joinChannels, "joinChannels is null").size() == 1, "Only single channel joins are supported");
+        int joinChannel = joinChannels.get(0);
+        this.addresses = requireNonNull(addresses, "addresses is null");
+        this.pagesHashStrategy = requireNonNull(pagesHashStrategy, "pagesHashStrategy is null");
+        this.channelCount = pagesHashStrategy.getChannelCount();
+
+        // reserve memory for the arrays
+        int hashSize = HashCommon.arraySize(addresses.size(), 0.75f);
+
+        // This implementation assumes arrays used in the hash are always a power of 2
+        mask = hashSize - 1;
+        key = new int[hashSize];
+        Arrays.fill(key, -1);
+
+        values = new long[addresses.size()];
+
+        // First extract all values from blocks to native array.
+        // Somehow having this as a separate loop is much faster compared
+        // to extracting hashes on the fly in the loop below..
+        extractValues(addresses, channels.get(joinChannel));
```

Comment:
I don't agree. This comment is tightly related not to this method, but the placement of this method within constructor. Especially that this comment is explicitly referencing next loop (ugh, I know it's ugly...)